MAID-2827: fix graceful utp shutdown

povilasb opened the pull request:
When Crust connects to remote peer, it does multiple connections in parallel. It only takes the first successful connection, the others were just dropped. That caused some problems with uTP connections since they should be closed gracefully with UtpStream::finalize(). That's what this PR does.

maidsafe-highfive commented:
Thanks for the pull request, and welcome! The MaidSafe team is excited to review your changes, and you should hear from @canndrew (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [CONTRIBUTOR.md](../blob/master/CONTRIBUTOR) for more information.

povilasb committed:
feat/pastream: update finalize() to gracefully close uTP connection

povilasb committed:
feat/connect: gracefully shutdown unused connections

When Crust connects to remote peer, it does multiple connections in parallel. It only takes the first successful connection, the others were just dropped. That caused some problems with uTP connections since they should be closed gracefully with UtpStream::finalize(). That's what this commit does.

povilasb committed:
fix/connect: remove connection that receives choose message from the pending connections list

Otherwise this connection gets polled twice: this connection is returned as first chosen connection and Crust attempts to call finalize() on this connection. That results in panic.

povilasb committed:
refactor/connect: extract some code to separate module

Makes it easier to read/modify code, etc.
